### PR TITLE
Support for favicons webpack plugin  + theme setup tools 

### PR DESCRIPTION
**Pull request by vesameskanen**
The PR is done - it:

- [ ] follows the style and naming rules (passes `npm run lint`)
- [ ] doesn't break anything (passes `npm run test-local` and `npm run test-browserstack`)
- [ ] design is as expected. NOTE! visuals are compared using HSL theme (`CONFIG=hsl npm run dev`).
-- If no design changes: `BS_USERNAME=user BS_ACCESS_KEY=key npm run test-visual` passes
-- If design changes: `BS_USERNAME=user BS_ACCESS_KEY=key npm run test-visual-update` to generate new images
- [ ] any changed files are transformed to ES6
- [ ] all new strings visible to users are

  * [ ] added to translations file
  * [ ] are translated to English, Finnish and Swedish (if not, add translations-needed label to PR)
  
- [ ] all changed components

  * [ ] have examples
  * [ ] are included in the style guide
  * [ ] are included in the visual tests (added to gemini tests)

If this PR fixes a bug, it includes a new test that catches the bug to prevent regressions.

### CODE REVIEW

File: app/component/AppBarLarge.js [hunk 1]
```diff
@@ -16,10 +16,18 @@ const AppBarLarge = ({ titleClicked }, { router, location, config }) => {
       },
     });
   };
+
   return (
     <div>
       <div className="top-bar bp-large flex-horizontal">
-        <a onClick={titleClicked} ><div className="navi-logo" /></a>
+        <a onClick={titleClicked} >
+          <section className="title">
```
Comment (hannesj) [inline app/component/AppBarLarge.js]: What is this used for?

File: app/configurations/config.hsl.js [hunk 2]
```diff
@@ -14,14 +14,19 @@ export default {
     FONT: 'https://cloud.typography.com/6364294/6653152/css/fonts.css',
   },
 
-  title: 'Reittiopas',
```
Comment (hannesj) [inline app/configurations/config.hsl.js]: This shouldn't be removed?

File: app/configurations/config.waltti.js [hunk 3]
```diff
@@ -0,0 +1,63 @@
+const API_URL = process.env.API_URL || 'https://dev-api.digitransit.fi';
+const APP_DESCRIPTION =
+  'Reittiopas uudistuu. Tule mukaan! Ota uuden uuden sukupolven matkaopas käyttöösi.';
```
Comment (hannesj) [inline app/configurations/config.waltti.js]: This should probably be something else?

File: app/configurations/config.oulu.js [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+  colors: {
+    primary: '#c40065',
+  },
+
+  socialMedia: {
+    title: 'Uusi Reittiopas - Oulu',
+    description: APP_DESCRIPTION,
+  },
+
+  title: 'Oulu Digitransit',
```
Comment (hannesj) [inline app/configurations/config.oulu.js]: The name Digitransit, should be not visible to customers, we should check all configs for this.

File: app/config.js [hunk 5]
```diff
@@ -1,48 +1,111 @@
...
+}
+
 function customizer(objValue, srcValue) {
   if (Array.isArray(objValue)) { return srcValue; } // Return only latest if array
   return undefined; // Otherwise use default customizer
 }
 
-export function getNamedConfiguration(configName) {
-  if (!configs[configName]) {
+export function addMetaData(config) {
```
Comment (hannesj) [inline app/config.js]: Should we add cache to the calculated metadata?

Comment (hannesj) [inline app/config.js]: Or do this while calculating the rest of the config

File: app/meta.js [hunk 6]
```diff
@@ -93,43 +70,21 @@ export default function getMetadata(lang, host, url, config) {
...
-      href: `${root}/img/${configPath}-icons/favicon-96x96.png`,
-    }, {
-      rel: 'icon',
-      type: 'image/png',
-      sizes: '16x16',
-      href: `${root}/img/${configPath}-icons/favicon-16x16.png`,
-    }, {
       rel: 'manifest',
-      href: `${root}/manifest.${configPath}.json`,
+      href: `${iconPath}manifest.json`,
```
Comment (hannesj) [inline app/meta.js]: these should include root i.e. `${root}/${iconPath}manifest.json`

File: app/config.js [hunk 7]
```diff
@@ -1,48 +1,111 @@
...
+  const html = stats.html.join(' ');
+
+  htmlParser.convert_html_to_json(html, (err, data) => {
+    if (!err) {
+      data.meta.forEach((e) => {
+        // eslint-disable-next-line no-param-reassign
+        delete e.innerHTML;
+        if (e.name === 'msapplication-config' || e.name === 'msapplication-TileImage') {
+          // eslint-disable-next-line no-param-reassign
+          e.content = `${stats.outputFilePrefix}${e.content}`; // fix path bug
```
Comment (hannesj) [inline app/config.js]: this should probably include config.APP_PATH

File: build/add-theme.js [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+  dir + '/_icon.scss',
+  `@import '../default/icon';`
+);
+
+fs.writeFileSync(dir + '/_theme.scss',
+`@import '../default/theme';
+
+/* main theme colors */
+$primary-color: ${color};
+$secondary-color: #008bde;
```
Comment (hannesj) [inline build/add-theme.js]: What about `darken($primary-color, 20%)`

File: app/server.js [hunk 9]
```diff
@@ -259,6 +260,8 @@ const isRobotRequest = agent =>
 
 export default function (req, res, next) {
   const config = getConfiguration(req);
+  addMetaData(config); // add dynamic metadata content
+  const application = appCreator(config);
```
Comment (hannesj) [inline app/server.js]: Application is already defined below https://github.com/HSLdevcom/digitransit-ui/pull/1275/files#diff-f53e59f241bb62f0261e245350683e07R279

File: app/configurations/config.waltti.js [hunk 10]
```diff
@@ -0,0 +1,63 @@
+const API_URL = process.env.API_URL || 'https://dev-api.digitransit.fi';
```
Comment (hannesj) [inline app/configurations/config.waltti.js]: Should we rename this file?

File: webpack.config.js [hunk 11]
```diff
@@ -93,15 +94,61 @@ function getRulesConfig(env) {
...
+}
+
+function getAllPossibleLanguages() {
+  return getAllConfigs().map(config => config.availableLanguages)
     .reduce((languages, languages2) => languages.concat(languages2))
     .filter((language, position, languages) => languages.indexOf(language) === position);
 }
 
+function faviconPluginFromConfig(config) {
+  let logo = './static/img/' + config.CONFIG + '-icon.png';
```
Comment (hannesj) [inline webpack.config.js]: Should we move the icons from static to under sass, so they are not copied for runtime?

File: app/configurations/config.hsl.js (hunk 2 above)
Comment (vesameskanen) [inline app/configurations/config.hsl.js]: Rebasing error, fixed.

File: app/configurations/config.waltti.js (hunk 3 above)
Comment (vesameskanen) [inline app/configurations/config.waltti.js]: Changed to 'Uusi Reittiopas'. Local per city configs can overrule this default.

File: app/configurations/config.oulu.js (hunk 4 above)
Comment (vesameskanen) [inline app/configurations/config.oulu.js]: Fixed in existing configs and config template.

File: webpack.config.js [hunk 12]
```diff
@@ -93,15 +94,61 @@ function getRulesConfig(env) {
   ]);
 }
 
-function getAllPossibleLanguages() {
+function getAllConfigs() {
   const srcDirectory = 'app/configurations';
   return fs.readdirSync(srcDirectory)
     .filter(file => /^config\.\w+\.js$/.test(file))
-    .map(file => require('./' + srcDirectory + '/' + file).default.availableLanguages)
+    .map(file => require('./' + srcDirectory + '/' + file).default);
```
Comment (hannesj) [inline webpack.config.js]: This should use the config merging from config.js

File: app/config.js (hunk 5 above)
Comment (vesameskanen) [inline app/config.js]: Good point. There is html parsing and all kinds of actions involved. Now part of config caching as you suggested

File: app/meta.js (hunk 6 above)
Comment (vesameskanen) [inline app/meta.js]: path definition already does this: const iconPath = `${root}/${path}/`;

File: app/config.js (hunk 7 above)
Comment (vesameskanen) [inline app/config.js]: Right, added.

File: build/add-theme.js (hunk 8 above)
Comment (vesameskanen) [inline build/add-theme.js]: Much better that way, done.

File: app/server.js (hunk 9 above)
Comment (vesameskanen) [inline app/server.js]: Fixed.

File: app/configurations/config.waltti.js (hunk 10 above)
Comment (vesameskanen) [inline app/configurations/config.waltti.js]: Renamed.

File: webpack.config.js (hunk 12 above)
Comment (vesameskanen) [inline webpack.config.js]: Well spotted - another piece of code that got lost in rebase. Fixed now.

File: webpack.config.js (hunk 11 above)
Comment (vesameskanen) [inline webpack.config.js]: makes sense. i'll move them while updating the svg icons you posted.

File: app/component/AppBarLarge.js (hunk 1 above)
Comment (vesameskanen) [inline app/component/AppBarLarge.js]: Just too much copypasting, fixed.

**Pull request merged**